pkg/services/images_build/main.go

```
@@ -115,16 +115,39 @@ func main() {
 					// route to specific event handler based on the event key
 					logevent.Debug("consumer routing based on key")
 
-					// execute the event handler if the record key has been defined
-					if _, exists := images.RegisteredEvents[key]; exists {
-						edgeEvent := images.RegisteredEvents[key]
-						json.Unmarshal(e.Value, edgeEvent)
-						// using reflection to avoid the compiler error with Consume() and an unknown struct
-						reflect.ValueOf(edgeEvent).MethodByName("Consume").Call(nil)
-					} else {
-						logevent.Warning("Skipping event. Record key is not defined")
+					switch key {
+					case models.EventTypeEdgeImageRequested:
+						// TODO: this seems like a lot of work to read an event back in
+						//			make this configurable and reusable
+						// unmarshal the event bytes to a specific struct
+						crcEvent := &images.ImageRequestedEvent{}
+						err = json.Unmarshal(e.Value, crcEvent)
+						if err != nil {
+							logevent.Error("Failed to unmarshal CRC event")
+						}
+
+						// CRC Event standard inlines the payload in Data field
+						//	we have to do some cast magic, but...
+						//	because of how Unmarshal handles interfaces, it's easier to go to bytes and back
+						edgePayloadString, err := json.Marshal(crcEvent.Data)
+						if err != nil {
+							logevent.Error("Failed to marshal Payload")
+						}
+						edgePayload := &models.EdgeImageRequestedEventPayload{}
+						err = json.Unmarshal(edgePayloadString, edgePayload)
+						if err != nil {
+							logevent.Error("Failed to unmarshal Payload")
+						}
+						// add the correct struct implementation for payload back to CRC struct in Data
+						crcEvent.Data = *edgePayload
+
+						// call the event's Consume method
+						go crcEvent.Consume()
+					default:
+						logevent.Warning("Record key is not recognized by consumer")
```

Are we sharing a topic? If so, this default may be called a lot and we should just silently continue, or change this log to Debug()

---

Switched it to trace while building these out. Each consumer will have its own consumer group. Do we need a way to see that an intended consumer group is not handling an event?